### PR TITLE
Fix #1509 - enable custom metrics introduced by sonar-cxx

### LINKED ISSUE #1509: sonar-cxx custom metrics are not dispayed/saved
1. go to `https://<server>/project/activity?id=<project_id>`
2. see metrics diagram -> choose `Custom` in the drop-down menu
3. click `Add metric` button
4. select one of the custom sonar-cxx metrics (see [CxxMetrics.java](https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxMetrics.java)) e.g. `CppCheck issues`
5. the diagram will show nothing

Plugin version: HEAD
SonarQube version: 7.1.0.11001

I guess, that the `CxxLanguage::MetricsCache` is not functioning properly. Is it really filled with data? Is it really persisted at the end?

### PR DESCRIPTION
**Pull request by ivangalkin**
close #1509 

Redesign of language-specific metrics
The following problems were identified and solved:

PROBLEM 1
* `CxxLanguage::MetricsCache` was not filled while construction of CxxLanguage object but only in `  CxxSquidSensor::CxxSquidSensor()`

SOLUTION
* Introduce `CxxMetricsFactory`.
* It has no dependencies to sensors.
* It creates language-dependent metrics in the constructor of `CxxSquidSensor`

PROBLEM 2
* During the test MetricsCache was almost always empty. In order to prevent warnings (?) missing metrics' lookups were just ignored

SOLUTON
* Introduce `protected abstract Optional<CxxMetricsFactory.Key> CxxReportSensor::getMetricKey();`
* By means of Optional<> semantic sensors show if metric is required or not
* CxxLanguage::getMetric() checks thows an exception if metric hasn't been found

PROBLEM 3
* CxxLanguage::getMetric(<required metric key>) expected the key to be typed as String
* In order to fill the MetricsCache the format {METRIC_KEY} was used
* Cache lookups were done however in the format {LANG_PROP_KEY}-{METRIC_KEY}

SOLUTION
* In order to make the lookup key type-safe and less error-prone I introduced the type   `CxxMetricsFactory.Key`

PROBLEM 4a

The plugin-specific metrics for public API were missing on module level

All CoreMetrics (incls. former Public API) have automatic aggregation:
1. ComputeEngine implements some server-side task
2. This task iterates over the project hierarchy
 `FILE -> DIRECTORY/MODULE -> ... -> ... -> PROJECT`
 and performs a recursive sum for all metrics raised
 on children component
3. This sum is stored as aggregated metric for the parent component
4. The number of hierarchy levels can be arbitrary

5. Starting with version 6.2 the public API metrics were
 deprecated and Sonar-CXX introduces its own plugin-specific
 public API metrics
6. These custom metrics are not aggregated automatically
 anymore

PROBLEM 4b

* PR #1398 introduced 9 new complexity metrics
They supported the aggregation of the file-level metrics.
The aggregation was implemeted inside of the `AstVisitor`,
however the aggregation was done for the parent module only.

* Ths solution worked for the simple SonarQube projects only.
Due to hierarchy depth == 2, aggregated results were stored
on the project level. Multi-module projects however might
have an arbitrary depth. So there was no proper aggregation
for such projects.

SOLUTION
* Unify implementation of custom metrics for Public-API and complexity
* Implement 2 Compute Engine tasks (see `MeasureComputer`), which perform
the aggregation and computation of these metrics
* Computaton/aggregation works for both types of SonarQube project -
the simple and the multi-module ones

PROBLEM 5
* The metric `CoreMetrics.COGNITIVE_COMPLEXITY` was measured but not persisted

MISC
* Test has been added
* Minor refactoring and comments

---
This change is [Reviewable](https://reviewable.io/reviews/sonaropencommunity/sonar-cxx/1514)

### CODE REVIEW

File: cxx-sensors/src/test/java/org/sonar/cxx/sensors/cppcheck/CxxCppCheckSensorTest.java [hunk 1]
```diff
@@ -86,6 +89,19 @@ public void shouldReportProjectLevelViolationsV2() {
     CxxCppCheckSensor sensor = new CxxCppCheckSensor(language);
     sensor.execute(context);
     assertThat(context.allIssues()).hasSize(3);
+
+    // assert that all all issues were filed on the same component (on the module)
```
Comment (Bertk) [inline cxx-sensors/src/test/java/org/sonar/cxx/sensors/cppcheck/CxxCppCheckSensorTest.java]: Please create separate tests for the module test and the metric test.

File: cxx-squid/src/main/java/org/sonar/cxx/CxxMetricsFactory.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  public static Map<CxxMetricsFactory.Key, Metric<?>> generateMap(String langKey, String langPropertiesKey) {
+    final String domain = langKey.toUpperCase(Locale.ENGLISH);
+
+    final Map<CxxMetricsFactory.Key, Metric<?>> metrics = new HashMap<>();
+
+    final Metric<Integer> PUBLIC_API = new Metric.Builder(Key.PUBLIC_API_KEY.getLanguageSpecificKey(langPropertiesKey),
+        "Public API", Metric.ValueType.INT)
+        .setDescription("Public API")
+        .setDirection(Metric.DIRECTION_WORST)
+         .setQualitative(false)
```
Comment (Bertk) [inline cxx-squid/src/main/java/org/sonar/cxx/CxxMetricsFactory.java]: _needlessly boxes a boolean constant_ (fb-contrib:NAB_NEEDLESS_BOOLEAN_CONSTANT_CONVERSION)
3 times in this file

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/functioncomplexity/CxxFunctionComplexitySquidSensor.java [hunk 3]
```diff
@@ -52,9 +55,9 @@
 
   private int linesOfCodeOverThreshold;
 
-  private Hashtable<SourceFile, FunctionCount> complexFunctionsPerFile = new Hashtable<>();
```
Comment (Bertk) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/functioncomplexity/CxxFunctionComplexitySquidSensor.java]: Can you avoid the synchronized Hashtable and use a HashMap instead (squid:S1149) for complexFunctionsPerFile , locInComplexFunctionsPerFile ?

Comment (ivangalkin) [reply cxx-sensors/src/main/java/org/sonar/cxx/sensors/functioncomplexity/CxxFunctionComplexitySquidSensor.java]: for now `org.sonar.squidbridge.AstScanner::scanFiles()` scans all files sequentially so synchronization is not necessary indeed. I will change Hashtable to HashMap, but whoever implements  #1493 must be aware of this change.

**Comment by ivangalkin** [comment]
Hi All,

this PR works perfectly for simple projects, but still doesn't support the multi-module projects. Also I am unhappy with metrics' aggregation though the custom introduced `Map`s. At least the density metrics (e.g. density of documented APIs, density of large functions and density of complex functions) can be easily calculated by means of [MeasureComputer](https://github.com/SonarSource/sonarqube/blob/master/sonar-plugin-api/src/main/java/org/sonar/api/ce/measure/MeasureComputer.java). The aggregation metrics (e.g. the number of public APIs for module = sum of public APIs of sub-modules) do not work like that unfortunately.

I would like to merge this PR and submit the solution for multi-module projects later.

@guwirth do you have any objections?

Regards

**Comment by ivangalkin** [comment]
hello @guwirth , @Bertk 
hello All,

I figured out, how to perform a depth-independent metrics aggregations. By means of ComputeEngine and the [MeasureComputer](https://github.com/SonarSource/sonarqube/blob/master/sonar-plugin-api/src/main/java/org/sonar/api/ce/measure/MeasureComputer.java) tasks one can propagate the measurement results across the whole project hierarchy. This technique works well for both, simple and multi-module SonarQube projects.

I've almost completely rewritten the custom metrics handling: there is no hand-made aggregation for complexity anymore, creation and calculation of all metrics (incl. Public API) are done in the standard way. Please review.

@ericlemes could you please also take a look? Thank you.

Please find below a short explanation w.r.t. aggregation. Please see the PU message also.

**PROBLEM 4a**
The plugin-specific metrics for public API were missing on module level

All CoreMetrics (incls. former Public API) have automatic aggregation:
1. ComputeEngine implements some server-side task
2. This task iterates over the project hierarchy
 `FILE -> DIRECTORY/MODULE -> ... -> ... -> PROJECT`
 and performs a recursive sum for all metrics raised
 on children component
3. This sum is stored as aggregated metric for the parent component
4. The number of hierarchy levels can be arbitrary

5. Starting with version 6.2 the public API metrics were
 deprecated and Sonar-CXX introduces its own plugin-specific
 public API metrics
6. These custom metrics are not aggregated automatically
 anymore

**PROBLEM 4b**
* PR #1398 introduced 9 new complexity metrics
They supported the aggregation of the file-level metrics.
The aggregation was implemeted inside of the `AstVisitor`,
however the aggregation was done for the parent module only.

Ths solution worked for the simple SonarQube projects only.
Due to hierarchy depth == 2, aggregated results were stored
on the project level. Multi-module projects however might
have an arbitrary depth. So there was no proper aggregation
for such projects.

**SOLUTION**
* Unify implementation of custom metrics for Public-API and complexity
* Implement 2 Compute Engine tasks (see `MeasureComputer`), which perform
the aggregation and computation of these metrics
* Computation/aggregation works for both types of SonarQube project -
the simple and the multi-module ones

**Pull request merged**